velox/substrait/VeloxToSubstraitExpr.cpp

```diff
@@ -207,6 +229,42 @@ VeloxToSubstraitExprConvertor::toSubstraitLiteral(
   return *literalExpr;
 }
 
+void VeloxToSubstraitExprConvertor::complexVectorToLiteral(
+    google::protobuf::Arena& arena,
+    std::shared_ptr<ConstantVector<ComplexType>> constantVector,
+    ::substrait::Expression_Literal* substraitLiteral) {
+  VELOX_CHECK_EQ(
+      constantVector->size(), 1, "Only one complex vector is expected.");
+  auto arrayVector = constantVector->valueVector()->as<ArrayVector>();
+  VELOX_CHECK_NOT_NULL(
+      arrayVector,
+      "arrayVector is NULL, Maybe Substrait literal is not supported for {}.",
+      constantVector->type()->toString());
+  VELOX_CHECK_EQ(arrayVector->size(), 1, "Only one array is expected.");
+  if (constantVector->isNullAt(0)) {
+    // Process the null value.
+    substraitLiteral->MergeFrom(exprConvertor_->toSubstraitNullLiteral(
+        arena, arrayVector->type()->kind()));
+  } else {
+    ::substrait::Expression_Literal_List* listLiteral =
+        google::protobuf::Arena::CreateMessage<
+            ::substrait::Expression_Literal_List>(&arena);
+    if (arrayVector->elements()->isScalar()) {
+      VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+          flatVectorToListLiteral,
+          arrayVector->elements()->type()->kind(),
+          arena,
+          arrayVector->elements(),
+          listLiteral);
+      substraitLiteral->mutable_list()->MergeFrom(*listLiteral);
+    } else {
+      VELOX_NYI(
+          "To Substrait literal is not supported for {}.",
```

Comment:
It is not clear to me how nested arrays are supported. Would you clarify?

Reply:
In fact, nested arrays will be parse to this way : "array_constructor(5 elements starting at 0 {0, 1, 2, 3, 4}".
I try a sql like "SELECT array[array[array[0, 1, 2, 3, 4]]]", and the projeciton expr is
 (gdb) p projections[0]->toString()
$3 = "array_constructor(array_constructor(5 elements starting at 0 {0, 1, 2, 3, 4}))" .
Only  innermost array will be parsed to ConstantTypedExpr.

Reply:
I see. PlanBuilder is just a test utility. As you pointed out it may not produce constant expressions with nested arrays. However, other plan generator can do that. Hence, we need to support and test these. Perhaps, you can use PlanBuilder::addNode to add a ProjectNode with a constant expression over nested array.

Reply:
I will use PlanBuilder::addNode to construct nested array.

Reply:
I add test for "SELECT array[], array[1, 2, null]." in this PR but no nested array.  Because I found a issue when I use addNode to construct nested array.  Function variantArrayToVector only work for scalar Type.  Could you give me some suggestions?
Here is my code.
`auto varArray = variant::array({variant::array({1, 2})});  `
` auto constantVector = BaseVector::wrapInConstant( 
     1, 0 , 
     variantArrayToVector(varArray.array(), pool_.get()));`
The good news is I am sure this UT  test can cover the newly added code.  The way I use "project(array[])"before can not cover.

Reply:
One option is using makeNullableNestedArrayVector method from the base class. Another option is using VectorFuzzer.

```
VectorFuzzer::Options opts;
opts.vectorSize = 100;

VectorFuzzer fuzzer;

fuzzer.fuzzConstant(ARRAY(ARRAY(BIGINT())));
```

Reply:
Thank you, I will try it.

Reply:
One way is to save the vector into DuckDB:

```
auto data = fuzzer.fuzzConstant(ARRAY(ARRAY(BIGINT())));
createDuckDbTable({data});
auto plan = ...
assertPlanConversion(plan, "SELECT * tmp");
```

Reply:
Sorry, This way is not useful.
`auto data = fuzzer.fuzzConstant(ARRAY(ARRAY(BIGINT())));
  createDuckDbTable({makeRowVector({data})});
`
DuckDB can't create table with array.

E1031 12:10:04.741909 732405 Exceptions.h:68] Line: ../../velox/exec/tests/utils/QueryAssertions.cpp:541, Function:createTable, Expression:  Parser Error: syntax error at or near "ARRAY"
LINE 1: CREATE TABLE tmp(c0 ARRAY[])